lib/util/progress.js

```diff
@@ -0,0 +1,145 @@
+'use babel'
+/** @jsx etch.dom */
+
+import etch from 'etch';
+
+import { view, Progress } from './etch';
+import Tooltip from './tooltip';
+
+// Progress Bars
+//
+// This module provides an API for displaying progress bars in Atom. The methods
+// below allow modifing the stack of progress bars, which is represented by
+// corresponding UI elements:
+//    - A "global" progress bar is shown in the status bar. In pratice, this is
+//      the first determinate progress bar in the stack. If there is none, the
+//      first element (which then is indeterminate) will be shown instead. If the
+//      stack is empty, an empty progress bar will be shown instead.
+//    - Hovering over that status bar element will show the complete stack as
+//      an overlay.
+//    - Hovering over the actual progress bar of one such stack element will show
+//      the message correesponding to that progress bar, if there is any.
+//
+// Methods:
+//
+// add(p = {progress: 0})
+//    Create and return a ProgressBar with the initial properties specified by `p`,
+//    which has the following methods available to it:
+//
+//    p.level
+//        Updates `p`s progress. `prog` is a number between 0 and 1 or `null`; if
+//        it is `null`, an indeterminate progress bar will be displayed.
+//
+//    p.leftText, p.rightText, p.message
+//        Sets the text displayed to the left of the progress bar, right of the
+//        progress bar, or when hovering over it.
+//
+//    p.destroy()
+//        Destroys `p` and removes it from the display stack.
+
+function clamp(x, min, max) {
+  return Math.min(Math.max(x, min), max);
+}
+
+class ProgressBar {
+  constructor(level, {leftText}={}) {
+    this._level = level;
+    this._leftText = leftText;
+  }
+  destroy() {
```

Comment:
It's a bit unfortunate that this refers to the global `stack`, since a `ProgressBar` could (in theory) now be created independently from everything else going on here. 
Maybe we could have `destroy` be the counterpart to `add` in the sense that both are *not* methods of `ProgressBar`, but rather of the implicit global class or something like that? Makes sense imho especially since this method doesn't reference it's object at all. 
Or, alternatively, just add this method after initialising the `ProgressBar` in `add`, which wouldn't change the API and seems much nicer than the first possibilty, now that I think about it ;)